### ISSUE TITLE
[@types/webpack-dev-middleware@4.1.0] still imports [@types/webpack] but those are legacy & incompatible.

### ISSUE THREAD

**Issue by rvion** [DefinitelyTyped/DefinitelyTyped#50773]
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I understand that `@types` package versions [do not follow SemVer](https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library).
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bumbleblym @reduckted @chrisabrams @peterblazejewicz 

## problem 

🔴 

`@types/webpack-dev-middleware@4.1.0`
imports `@types/webpack` 
but those are legacy & incompatible.
`webpack` now ships with its own typings

## infos

see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50385#issuecomment-765413654

**Comment by rvion** [comment]
https://webpack.js.org/blog/2020-10-10-webpack-5-release/#typescript-typings

**Comment by rvion** [comment]
it also seems that there are also 2 more issues: 

`webpack.Watching`
`webpack.Stats.ToJsonOptions;`

do not typecheck

**Comment by tqma113** [comment]
I will fix it. Thanks for correcting.

**Comment by tqma113** [comment]
Thanks for issue. I will work on it.

**Comment by tqma113** [comment]
`@types/webpack`: I find a lot of packages depend on `webpack` have this issue, like `webpack-dev-server`, `webpack-hot-middleware`, `webpack-manifest-plugin` and so on. They also have the dependence `@types/webpack` in `package.json`. And I have searched issue but did not find the answer. But I find then `terser-webpack-plugin` does not have this problem. It set the dependence `"webpack": "5.1.0".`https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b71dc542f83c8c655713a7c9abedcdfb62dfa845/types/terser-webpack-plugin/package.json#L5
Maybe this is the correct way.

**Comment by tqma113** [comment]
Checking for `webpack.Stats.ToJsonOptions`: In the latest version of webpack, the type of parameter of `stats.toJson` is `any`
https://github.com/webpack/webpack/blob/d02e8daa267ca094fc0a030a595b000d35a99493/types.d.ts#L9605
So at present there is no type checking. And the type `webpack.Stats.ToJsonOptions` in this `@types/webpack-dev-middleware`, maybe we need fix it(but not very necessary, maybe there will be some type in next version `webpack`).

**Comment by tqma113** [comment]
Checking for `webpack.Watching`: This also need to fix. Maybe we should fix it with `webpack.Compiler["watching"]`.

**Comment by tqma113** [comment]
The `@types/webpack` problem need to discuss. Can you give some help? @peterblazejewicz

**Comment by rvion** [comment]
> `@types/webpack`: I find a lot of packages depend on `webpack` have this issue, like `webpack-dev-server`, `webpack-hot-middleware`, `webpack-manifest-plugin` and so on. 

Yes, when I saw that, I forked those packages locally to remove the dependency complely.

misc pointers in case it helps: 

- `webpack-hot-middleware` has a webpack-5 compatible (4.1.0) version with almost correct types beside the extra @types/webpack import (https://www.npmjs.com/package/@types/webpack-dev-middleware)

- `webpack-dev-server` has a `v4.0.0-beta.0` version that is compatible with `webpack 5`, but has no corresponding @types/... 
https://www.npmjs.com/package/@types/webpack-dev-middleware

--------------------

> It set the dependence `"webpack": "5.1.0".`
> Maybe this is the correct way.

🟢 I believe this is a possible correct way. 
You could also add it as devDependency or peerDependency, or not add it at all depending on various assumptions you want to do about people who install it.

**Comment by tqma113** [comment]
`terser-webpack-plugin`'s way  is not correct. When I used it I got the type conflect error.
<details>
<summary>
Error log
</summary>
<pre>
<code>
Type 'TerserPlugin' is not assignable to type '"..." | WebpackPluginInstance | ((this: Compiler, compiler: Compiler) => void)'.
  Type 'TerserPlugin' is not assignable to type 'WebpackPluginInstance'.
    Types of property 'apply' are incompatible.
      Type '(compiler: import("/Users/tqma/Projects/github/react-torch/node_modules/@types/terser-webpack-plugin/node_modules/webpack/types").Compiler) => void' is not assignable to type '(compiler: import("/Users/tqma/Projects/github/react-torch/node_modules/webpack/types").Compiler) => void'.
        Types of parameters 'compiler' and 'compiler' are incompatible.
          Type 'import("/Users/tqma/Projects/github/react-torch/node_modules/webpack/types").Compiler' is not assignable to type 'import("/Users/tqma/Projects/github/react-torch/node_modules/@types/terser-webpack-plugin/node_modules/webpack/types").Compiler'.
            The types of 'hooks.shouldEmit' are incompatible between these types.
              Type 'import("/Users/tqma/Projects/github/react-torch/node_modules/tapable/tapable").SyncBailHook<[import("/Users/tqma/Projects/github/react-torch/node_modules/webpack/types").Compilation], boolean, import("/Users/tqma/Projects/github/react-torch/node_modules/tapable/tapable").UnsetAdditionalOptions>' is not assignable to type 'import("/Users/tqma/Projects/github/react-torch/node_modules/tapable/tapable").SyncBailHook<[import("/Users/tqma/Projects/github/react-torch/node_modules/@types/terser-webpack-plugin/node_modules/webpack/types").Compilation], boolean, import("/Users/tqma/Projects/github/react-torch/node_modules/tapable/tapable").Uns...'.

 54         new TerserPlugin({
            ~~~~~~~~~~~~~~~~~~
 55           terserOptions: {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
... 
 84           parallel: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~
 85         }),
    ~~~~~~~~~~


Found 1 error.

error Command failed with exit code 2.
</code>
</pre>
</details>

**Comment by rvion** [comment]
@tqma113 I think you misunderstood this. This issue is specifically about those conflict that arise because several packages still require conflicing legacy @types/webpack ? on a test npm project with nothing but webpack and terser, do you see any conflict ?

**Comment by tqma113** [comment]
If in a project with nothing but webpack and terser, there is no conflict. I methioned `terser-webpack-plugin` provide a solution before. After my testing, i find it is not perfect solution. Let's continue to try.

**Comment by rvion** [comment]
@tqma After my testing, I find it is the perfect solution, and it's just that other packages need to do the same. I also think it's probably the only correct solution.

**Comment by peterblazejewicz** [comment]
@tqma113 missed this one. There is no clean way to fix v4/v5 problems (unless something changed just recently). There is always a problem either here (on DT) or when consuming packages.

**Comment by tqma113** [comment]
I continue to pay attention to this issue. As you mentioned there is no way unless something change. So it is blocked. I think it needs someone to push this change. I can't do anything at the moment.

**Comment by unstubbable** [comment]
For everyone who is as impatient as me, there is a work-around using module augmentation to make the webpack 5 types compatible with the `webpack-dev-middleware` types (as well as `ts-loader`):

```ts
// local webpack.d.ts
import * as webpack from 'webpack';

declare module 'webpack' {
  namespace loader {
    // Stub to avoid compiler error in ts-loader types.
    interface LoaderContext {}
  }

  namespace Stats {
    // Work-around for https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50802
    type ToJsonOptions = Parameters<webpack.Stats['toJson']>[0];
  }
}
```

**Comment by peterblazejewicz** [comment]
oh, maybe there should be a DT shim package for webpack 4 <> 5 compatiblity, with approach like above augmentation? (@types/webpack-shim)

**Comment by rchl** [comment]
It seems like some solution was implemented for many webpack-related types in https://github.com/DefinitelyTyped/DefinitelyTyped/commit/a5056dc4a76e37785c52fe7abdbca80dcb80d91a#diff-6263d7ce12ba21c8cd463d989f0f45c3cb390c6525eb87a515d9ed7410779123

This package is one left that still depends on `"@types/webpack" "*"` while it should probably be depending on `"@types/webpack" "^4"`.

I don't fully understand the solution but my naive thinking is that the 4.x version of this type package should depend on `^4` while the 5.x version should depend on webpack 5 (or `*`).

@sandersn

**Comment by sandersn** [comment]
Yes, that's what should happen now that `@types/webpack@5` is a passthrough for `webpack@5`'s types. For older versions of this package, you can make `paths` redirects in tsconfig to older versions of webpack.

**Comment by rchl** [comment]
@tqma113 I see that you've created a PR to switch to explicitly using webpack 5 types for the latest version of this package.

Will you also separately update version 4 of this package? It needs to point to webpack 4 types specifically as currently, it does point to `*` which, depending on the project, can pick up webpack 5 types.

**Comment by tqma113** [comment]
@rchl This is indeed a problem. `webpack-dev-middleware` is indeed compatible with `webpack@4`. Maybe we need to add one `v4` to `@types/webpack-dev-middleware`. @peterblazejewicz  Can you give some suggestions?

**Comment by rchl** [comment]
Actually, this was already fixed in the 4.x line: https://unpkg.com/%40types%2Fwebpack-dev-middleware@4.1.2/package.json

**Comment by rchl** [comment]
So it works for me now because I'm using webpack 4 but I'm not sure if it would work for those using webpack 5.

The actual [webpack-dev-middleware](https://github.com/webpack/webpack-dev-middleware/releases) seems to be compatible with both webpack 4 and 5 so I'm not sure how that should be handled from the perspective of the types. Maybe there should be a v5 release of types added while keeping v4 dependent on v4 webpack types? I don't fully understand how versioning works with types.

But I hope that your PR won't break it for webpack 4 users again :)

**Comment by feebabs** [comment]
I am running into this same issue with `@types/zip-webpack-plugin`. Looking through the package, it has a dependency on `"@types/webpack": "^4"` which is causing issues with our repo since we are using webpack 5 which includes it's own types. While trying to build, I am getting the error:

```
[webpack-cli] configuration/webpack/base.config.ts(38,5): error TS2322: Type 'ZipPlugin' is not assignable to type 'WebpackPluginInstance | ((this: Compiler, compiler: Compiler) => void)'.
  Type 'ZipPlugin' is not assignable to type 'WebpackPluginInstance'.
    Types of property 'apply' are incompatible.
      Type '(compiler: import("REDACTED/node_modules/@types/zip-webpack-plugin/node_modules/@types/webpack/index").Compiler) => void' is not assignable to type '(compiler: import("REDACTED/node_modules/webpack/types").Compiler) => void'.
        Types of parameters 'compiler' and 'compiler' are incompatible.
          Property '_pluginCompat' is missing in type 'import("REDACTED/node_modules/webpack/types").Compiler' but required in type 'import("REDACTED/node_modules/@types/zip-webpack-plugin/node_modules/@types/webpack/index").Compiler'.
```

I have tried a few different workarounds to get it to build to no avail. Short of lowering my webpack version back down to 4, is there anything we can do?
